server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    private SnapshotShardContext createDummyContext() {
+        return createDummyContext(new SnapshotId(randomAlphaOfLength(10), UUIDs.randomBase64UUID()));
+    }
+
+    private SnapshotShardContext createDummyContext(final SnapshotId snapshotId) {
+        IndexId indexId = new IndexId(randomAlphaOfLength(10), UUIDs.randomBase64UUID());
+        ShardId shardId = new ShardId(indexId.getName(), indexId.getId(), 1);
+        Settings settings = Settings.builder()
+            .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .build();
+        IndexSettings indexSettings = new IndexSettings(
+            IndexMetadata.builder(indexId.getName()).settings(settings).build(),
+            Settings.EMPTY
+        );
+        Store dummyStore = new Store(shardId, indexSettings, new ByteBuffersDirectory(), new DummyShardLock(shardId));
+        return new SnapshotShardContext(
+            dummyStore,
+            null,
+            snapshotId,
+            indexId,
+            new Engine.IndexCommitRef(null, () -> {}),
+            null,
+            IndexShardSnapshotStatus.newInitializing(null),
+            Version.CURRENT,
+            Collections.emptyMap(),
+            ActionListener.noop()
+        );
+    }
+
+    public void testEnqueueCreatesNewWorkersIfNecessary() throws Exception {
+        int maxSize = randomIntBetween(1, threadPool.info(ThreadPool.Names.SNAPSHOT).getMax());
+        CountDownLatch snapshotBlocker = new CountDownLatch(1);
+        MockedRepo repo = new MockedRepo(snapshotBlocker);
+        ShardSnapshotWorkers workers = new ShardSnapshotWorkers(maxSize, executor, repo::snapshotShard, repo::snapshotFile);
+        repo.setWorkers(workers);
+        int enqueuedSnapshots = maxSize - 1; // So that it is possible to create at least one more worker
+        for (int i = 0; i < enqueuedSnapshots; i++) {
+            workers.enqueueShardSnapshot(createDummyContext());
+        }
+        assertBusy(() -> assertThat(workers.size(), equalTo(maxSize - 1)));
+        // Adding at least one new shard snapshot would create a new worker
+        int newTasks = randomIntBetween(1, 10);
+        for (int i = 0; i < newTasks; i++) {
+            workers.enqueueShardSnapshot(createDummyContext());
+        }
+        enqueuedSnapshots += newTasks;
+        assertThat(workers.size(), equalTo(maxSize));
+        snapshotBlocker.countDown();
+        // Eventually all workers exit
+        assertBusy(() -> assertThat(workers.size(), equalTo(0)));
+        assertThat(repo.finishedFileSnapshotTasks(), equalTo(repo.expectedFileSnapshotTasks()));
+        assertThat(repo.finishedShardSnapshotTasks(), equalTo(enqueuedSnapshots));
+        assertThat(repo.finishedShardSnapshots(), equalTo(enqueuedSnapshots));
+    }
+
+    public void testCompareToShardSnapshotTask() {
```

Comment:
To test that ShardSnapshotTask has priority over file snapshots, I tried to add some tests beyond just this compare test, but it turned out quite tricky to verify that. If you have any suggestions, please let me know.